preview/MsixCore/msixmgr/AutoPlay.cpp

```diff
@@ -0,0 +1,577 @@
+    tempProgIDBuilder.append(tempLowerBuffer);
+
+    if (!subCategory.empty())
+    {
+        tempLowerBuffer.assign(subCategory);
+        std::transform(tempLowerBuffer.begin(), tempLowerBuffer.end(), tempLowerBuffer.begin(), ::tolower);
+        tempProgIDBuilder.append(tempLowerBuffer);
+    }
+
+    // Create the crypto provider and start the digest / hash
+    AutoPtr<CryptoProvider> cryptoProvider;
+    RETURN_IF_FAILED(CryptoProvider::Create(&cryptoProvider));
+    RETURN_IF_FAILED(cryptoProvider->StartDigest());
+    COMMON_BYTES data = { 0 };
+    data.length = (ULONG)tempProgIDBuilder.size() * sizeof(WCHAR);
+    data.bytes = (LPBYTE)tempProgIDBuilder.c_str();
+
+    RETURN_IF_FAILED(cryptoProvider->DigestData(&data));
+
+    // Grab the crypto digest
+    COMMON_BYTES digest = { 0 };
+    RETURN_IF_FAILED(cryptoProvider->GetDigest(&digest));
+
+    // Ensure the string buffer has enough capacity
+    StringBuffer base32EncodedDigest;
+    RETURN_IF_FAILED(base32EncodedDigest.SetCapacity(
+        MaxBase32EncodedStringLength));
+
+    // Base 32 encode the bytes of the digest and put them into the string buffer
+    ULONG base32EncodedDigestCharCount = 0;
+    RETURN_IF_FAILED(Base32Encoding::GetChars(
+        digest.bytes,
+        GetMinValue(digest.length, MaxByteCountOfDigest),
+        MaxBase32EncodedStringLength,
+        base32EncodedDigest.GetChars(),
+        &base32EncodedDigestCharCount));
+
+    // Set the length of the string buffer to the appropriate value
+    RETURN_IF_FAILED(base32EncodedDigest.SetLength(base32EncodedDigestCharCount));
+
+    // ProgID name is formed by appending the encoded digest to the "AppX" prefix string
+    tempProgIDBuilder.clear();
+
+    tempProgIDBuilder.append(AppXPrefix);
+    tempProgIDBuilder.append(base32EncodedDigest.GetString()->chars);
+
+    assert(tempProgIDBuilder.GetLength() <= MaxProgIDLength);
+
+    // Set the return value
+    generatedProgId.assign(tempProgIDBuilder.c_str());
+
+    return S_OK;
+}
+
+HRESULT AutoPlay::GenerateHandlerName(LPWSTR type, const std::wstring handlerNameSeed, std::wstring & generatedHandlerName)
+{
+    // Constants
+    static const ULONG HashedByteCount = 32;      // SHA256 generates 256 hashed bits, which is 32 bytes
+    static const ULONG Base32EncodedLength = 52;  // SHA256 generates 256 hashed bits, which is 52 characters after base 32 encoding (5 bits per character)
+
+    std::wstring packageFamilyMoniker = m_msixRequest->GetPackageInfo()->GetPackageFamilyName();
+    std::wstring applicationId = m_msixRequest->GetPackageInfo()->GetApplicationId();
+    std::wstring handlerNameBuilder;
+    HCRYPTPROV hProv = NULL;
+    HCRYPTHASH hHash = NULL;
+    DWORD hashLength;
+    BYTE bytes[HashedByteCount];
+    ULONG base32EncodedDigestCharCount;
+    StringBuffer base32EncodedDigest;
+    size_t typeLength;
+
+    // First, convert Package Moniker and App Id to StringBuffers for convenience - lowercase the values so that the comparison

```

Comment:
>StringBuffers [](start = 52, length = 13)

comment not accurate. you're using std::wstring.

Reply:
Updated comment to not use moniker and changed comment to match usage of std::wstring instead of string buffer.